extras/bme680/bme680.h

```diff
@@ -0,0 +1,351 @@
+ * measurement cycle, raw sensor data are available and the sensor returns
+ * automatically back to sleep mode.
+ *
+ * Using the BME680 consists of the following steps
+ *
+ * 1. Trigger the sensor to switch into forced mode to perform one THPG cycle
+ * 2. Wait until the THPG cycle has been finished (measurement duration)
+ * 3. Fetch raw sensor data, compensate and convert them to sensor values
+ *
+ * ---------------------------------------------------------------------------
+ */
+
+/**
+ * @brief	Initialize a BME680 sensor
+ * 
+ * The function initializes the sensor device data structure, probes the 
+ * sensor, soft resets the sensor, and configures the sensor with default
+ * settings: 
+ *    
+ * - Oversampling rate for temperature, pressure, humidity is osr_1x
+ * - Filter size for pressure and temperature is iir_size 3
+ * - Heater profile for gas is 320 degree Celsius for 150 milliseconds.
+ *
+ * The sensor can be connected either to an I2C or a SPI bus. In both cases,
+ * the parameter *bus* specifies the ID of the corresponding bus. Please note
+ * that in case of SPI, bus 1 has to be used since bus 0 is used for system
+ * flash memory.
+ *
+ * If parameter *addr* is greater than 0, it defines a valid I2C slave address
+ * and the sensor is connected to an I2C bus. In that case parameter *cs* is
+ * ignored.
+ *
+ * If parameter *addr* is 0, the sensor is connected to a SPI bus. In that
+ * case, parameter *cs* defines the GPIO used as CS signal 
+ *
+ * @param   bus     I2C or SPI bus at which BME680 sensor is connected
+ * @param   addr    I2C addr of the BME680 sensor, 0 for SPI
+ * @param   cs      SPI CS GPIO, ignored for I2C
+ * @return          pointer to sensor data structure, or NULL on error
+ */
+bme680_sensor_t* bme680_init_sensor (uint8_t bus, uint8_t addr, uint8_t cs_pin);
+
+
+/**
+ * @brief	Force one single TPHG measurement
+ *
+ * The function triggers the sensor to start one THPG measurement cycle in
+ * forced mode, the only measurement mode supported by the BME680. Measurement
+ * parameters for the measurement like oversampling rates, IIR filter sizes
+ * and heater profile can be configured before.
+ *
+ * On success, the function returns an estimated measurement duration given
+ * in RTOS ticks. This is the time in ticks needed by the sensor before
+ * measurement results become available. The user task has to wait this
+ * duration before it can use function *bme680_get_results_fixed* or
+ * function *bme680_get_results_float* to fetch the measurement results.
+ *
+ * The measurement duration strongly depends on which measurements in the THPG
+ * measurement cycle are performed and which configuration parameters were set.
+ * It can vary from 1 RTOS (10 ms) tick up to 4500 RTOS ticks (4.5 seconds).
+ *
+ * @param   dev   pointer to the sensor device data structure
+ * @return        measurement duration given in RTOS ticks or -1 on error
+ */
+int32_t bme680_force_measurement (bme680_sensor_t* dev);
```

Comment:
Could I suggest returning a time in milliseconds or would 1ms be far too long?

Reply:
I'm a bit unsure. Initially, this function returned the duration in milliseconds. It was the same in the SHT3x driver.

During the review of SHT3x driver, @ourairquality had the good idea to return the measurement duration simply in ticks so that they can be used directly as parameter of function vTaskDelay to wait. We had to realize that vTaskDelay(measurement_duration_in_ms+(portTICK_PERIOD_MS-1))/portTICK_PERIOD_MS) would result in a waiting time that is too short. Reason is that the first tick and in worst cases the first 2 ticks are shorter than portTICK_PERIOD_MS. To avoid problems at user task, we thought it is better to return a number of ticks that is high enough in any case.

Reply:
At the moment, I'm separating the function to start the measurement and a function to get the measurement duration. This results from the discussion with @ourairquality about the SHT3x driver. 

Since the measurement duration depends only on the configuration of the measurement cycle, it is not necessary to compute the duration every time a measurement is started. It seems to be enough to configure the sensor, to get the measurement duration for this configuration and then to start the measurement multiple times. Measurement duration has to be only recalculated when the measurement configuration is changed. That is, as long as the measurement configuration isn't changed, the measurement duration can be considered as constant. So I'm changing the interface in that way:

```
bme680_set_... // set measurement configuration
uint32_t duration = bme680_get_measurement_duration(...);

while (1)
{
    if (bme680_force_measurement(...)) 
    {
        vTaskDelay(duration);
        bme_680_get_results_float(...);
    }
    ...
}
```

Reply:
I see you have good reasons to return the duration in ticks which is fine by me.